apps/console/src/features/roles/components/edit-role/edit-role-groups-updated.tsx

```diff
@@ -0,0 +1,547 @@
+                return {
+                    displayName: group.display,
+                    id: group.value
+                };
+            }));
+        } else {
+            setInitialSelectedGroupsOptions([]);
+            setShowEmptyRolesListPlaceholder(true);
+        }
+    }, [ role ]);
+
+    /**
+     * set selected groups options initially
+     */
+    useEffect(() => {
+        if (!selectedGroupsOptions && initialSelectedGroupsOptions) {
+            setSelectedGroupsOptions(initialSelectedGroupsOptions);
+        }
+
+        // set groups options when the user is read only
+        if (isReadOnly && initialSelectedGroupsOptions) {
+            setGroupsOptions(initialSelectedGroupsOptions);
+        }
+    }, [ isReadOnly, initialSelectedGroupsOptions ]);
+
+    /**
+     * set groups options
+     */
+    useEffect(() => {
+        if (!isReadOnly) {
+            if (originalGroupList && originalGroupList?.totalResults !== 0) {
+                setGroupsOptions(originalGroupList?.Resources);
+            } else {
+                setGroupsOptions([]);
+            }
+        }
+    }, [ originalGroupList ]);
+
+    /**
+     * set removed groups
+     */
+    useEffect(() => {
+        if (!isReadOnly && initialSelectedGroupsOptions && selectedGroupsOptions) {
+            setRemovedGroupsOptions(initialSelectedGroupsOptions?.filter((group: GroupsInterface) => {
+                return selectedGroupsOptions?.find(
+                    (selectedGroup: GroupsInterface) => selectedGroup.id === group.id) === undefined;
+            }));
+        }
+    }, [ initialSelectedGroupsOptions, selectedGroupsOptions ]);
+    
+    /**
+     * call mutateGroupListFetchRequest when the group search value changes
+     */
+    useEffect(() => {
+        if ( groupSearchValue ) {
+            mutateGroupListFetchRequest().finally(() => {
+                setGroupSearchLoading(false);
+            });
+        }
+    }, [ groupSearchValue ]);
+    
+    /**
+     * show error if group list fetch request failed
+     */ 
+    useEffect(() => {
+        if ( groupListFetchRequestError ) {
+            dispatch(
+                addAlert({
+                    description: t("console:manage.features.roles.edit.groups.notifications.fetchError.description"),
+                    level: AlertLevels.ERROR,
+                    message: t("console:manage.features.roles.edit.groups.notifications.fetchError.message")
+                })
+            );
+        }
+    }, [ groupListFetchRequestError ]);
+
+    /**
+     * Handle the restore groups.
+     * 
+     * @param remainingGroups - remaining groups
+     */
+    const handleRestoreGroups = (remainingGroups: GroupsInterface[]) => {
+        const removedGroups: GroupsInterface[] = [];
+
+        removedGroupsOptions.forEach((group: GroupsInterface) => {
+            if (!remainingGroups.find((newGroup: GroupsInterface) => newGroup.id === group.id)) {
```

Comment:
Shall we add nullable checks?

Reply:
Fixed with https://github.com/wso2/identity-apps/pull/4212/commits/c57c7b953b186b80c95a389a9c46805020b584fa